stonesoup/types/particle.py

```diff
@@ -25,3 +29,71 @@ def __init__(self, state_vector, weight, parent=None, *args, **kwargs):
     @property
     def ndim(self):
         return self.state_vector.shape[0]
+
+
+class Particles(Type):
+    """
+    Particle type
+
+    A collection of particles. Contains a state and weight for each particle
+    """
+    state_vector: StateVectors = Property(default=None, doc="State vectors of particles")
+    weight: MutableSequence[Probability] = Property(default=None, doc='Weights of particles')
+    parent: 'Particles' = Property(default=None, doc='Parent particles')
+    particle_list: MutableSequence[Particle] = Property(default=None,
+                                                        doc='List of Particle objects')
+
+    def __init__(self, state_vector=None, weight=None, parent=None, particle_list=None,
+                 *args, **kwargs):
+        if state_vector is not None:
+            assert particle_list is None,\
+                "Particles object cannot use both state_vector and particle_list"
+
+        if particle_list and isinstance(particle_list, list):
+            state_vector = StateVectors([particle.state_vector for particle in particle_list])
+            weight = np.array([Probability(particle.weight) for particle in particle_list])
+            parent_list = [particle.parent for particle in particle_list
+                           if particle.parent is not None]
+            if parent_list:
+                parent = Particles(particle_list=parent_list)
+
+        if parent:
+            parent.parent = None
+
+        if state_vector is not None and not isinstance(state_vector, StateVectors):
+            state_vector = StateVectors(state_vector)
+        if weight is not None and isinstance(weight, np.ndarray):
+            weight = np.array(weight)
+
+        self._index = -1
+
+        super().__init__(state_vector, weight, parent, particle_list, *args, **kwargs)
+
+    def __getitem__(self, item):
+        if self.parent:
+            p = self.parent[item]
+        else:
+            p = None
+
+        particle = Particle(state_vector=self.state_vector[:, item],
+                            weight=self.weight[item],
+                            parent=p)
+        return particle
+
+    def __len__(self):
+        return self.state_vector.shape[1]
+
+    def __iter__(self):
+        return self
```

Comment:
Not sure we should make this class an iterator, but instead return a generator (so make the class iterable still). Making this an iterator, means that you can't iterate over this in parallel, due to having to use an `_index` variable. 

Something like:
```python
def __iter__(self):
    yield from (self[i] for i in range(len(particles))
```
and drop the `__next__` method and `_index`.

Reply:
This was my first attempt at making it also work just like a list of particles (for testing). In the end I added getitem, so I'll delete the iter and next methods.